Allow CustomDist with inferred logp in Mixture

ricardoV94 opened the pull request:
This PR changes the logic used for logprob inference. Instead of eager bottom-up conversion to measurable variables in the IR rewrites, we only convert nodes whose outputs were marked as "needs_measuring". This is achieved with the new `PreserveRVMappings.request_measurable` method.

This strategy obviates the need to undo unnecessary conversions. It also obviates a subtle need for graph cloning via the `ignore_logprob` helper, which prevented intermediate measurable rewrites from being reversed when they were needed to derive the logprob of valued variables, but were not directly valued. This indirect role of `ignore_logprob` is now done more explicitly and efficiently via the `request_measurable` method.

All other uses of `ignore_logprob` (and `reconsider_logprob`) were removed from the codebase

The `get_measurable_outputs` dispatching was also abandoned in favor of only considering outputs associated with value variables.

A new MergeOptimizerRewrite was written to further target local rewrites to only those nodes whose variables have been marked as `needs_measuring`. This should be rather more efficient

Closes #6728 
Supersedes #6698 

CC @shreyas3156 @Dhruvanshu-Joshi 

<!-- readthedocs-preview pymc start -->
----
:books: Documentation preview :books:: https://pymc--6746.org.readthedocs.build/en/6746/

<!-- readthedocs-preview pymc end -->

ricardoV94 committed:
New logprob inference logic

This commit changes the logic used for logprob inference. Instead of eager bottom-up conversion to measurable variables in the IR rewrites, we only convert nodes whose outputs were marked as "needs_measuring". This is achieved with the new `PreserveRVMappings.request_measurable` method.

This strategy obviates the need to undo unnecessary conversions. It also obviates a subtle need for graph cloning via the `ignore_logprob` helper, which prevented intermediate measurable rewrites from being reversed when they were needed to derive the logprob of valued variables, but were not directly valued. This indirect role of `ignore_logprob` is now done more explicitly and efficiently via the `request_measurable` method.

All other uses of `ignore_logprob` (and `reconsider_logprob`) were removed from the codebase

The `get_measurable_outputs` dispatching was also abandoned in favor of only considering outputs associated with value variables.

A new MergeOptimizerRewrite was written to further target local rewrites to only those nodes whose variables have been marked as `needs_measuring`.

ricardoV94 committed:
Cleanup RV in graph checks

ricardoV94 committed:
Fix logprob of `check_and_raise`

ricardoV94 commented on .github/workflows/tests.yml:
```diff
@@ -105,6 +105,7 @@ jobs:
             tests/logprob/test_abstract.py
             tests/logprob/test_basic.py
             tests/logprob/test_binary.py
+            tests/logprob/test_checks.py
```
This feels very dangerous, there is no fallback for tests that are not included explicitly?

These tests were completely broken for a while

Dhruvanshu-Joshi replied:
Previously, the asserts were not being defined in this test effectively and hence the desired exceptions were not raised. I think the current changes should be sufficient to rectify this.

ricardoV94 replied:
The problem I am referring to is not the underlying bugs, but the fact that we were not running the tests for this file at all, which is a more broad issue.

michaelosthege replied:
We have a script under `./scripts` that should have caught this. Why didn't it?

ricardoV94 replied:
No idea, can you take a look?

michaelosthege replied:
Log indicates that the check was skipped.
Maybe because this line doesn't match the filename of the `tests.yml` workflow: https://github.com/pymc-devs/pymc/blob/4d2b2efc2b8ad7b5060d3f4a642534408b5551eb/.pre-commit-config.yaml#L67


I'm just on my phone in the train, so I can't open a PR. Can you try it on this branch?

ricardoV94 replied:
Opened PR in #6756

ricardoV94 committed:
Allow composition of interdependent container variables

Join/MakeVector/IfElse can output multiple interdependent variables. These are potentially measurable because in the logp each output is given a distinct value variable. However, this isn't known during the IR rewrites.

To circumvent this issue, we run an inner IR rewrite after giving dummy value variables to each output

ricardoV94 committed:
Remove unused test util

ricardoV94 committed:
Fix nested and single output IfElse logp

ricardoV94 committed:
Fix auto-naming of multi-output logprobs
